CHANGELOG.md

```diff
@@ -10,6 +10,16 @@ how a consumer would use the library (e.g. adding unit tests, updating documenta
 
 ## Unreleased
 
+### Added
+
+-  `sdk.alertrules` methods:
+    - `add_user()`
```

Comment:
Method names needs to be rethought. 

Here `add_user` takes a custom identifier and multiple email aliases in the user alert rules. Hence `add_user` is not  appropriate.  `add_user_group` would be more appropriate. 
   i.e. add_user_group(rule_id, group_identifier, [email1, email2] 

Similarly for `remove_users` and `remove_all_users` , 

I was not happy with any of the names I could think of and hence have left it for discussion in review.

Reply:
Moreover there is one more endpoint in the API, which accepts rule_id, user_group_id and aliases, exactly like add_users and removes only the aliases from the rule. 

We might want to support that too, do we need a new method for that !?

Reply:
`add_user` refers to a single code42 user who could have have multiple email addresses used by other services. The `group_identifier` you mention is the code42 username, while `[email1, email2]` is a set of known _alternate_ email addresses that the user has that otherwise may be unknown to code42.

I _think_ that this list in practice should be the same as the list of cloud aliases from `detectionlists.users` -- I will verify this. If this is the case, we should change our implementation to simply look those up and not require that the user specifiy them. I will do some research into this and get back to you here with details.

Reply:
I did some research into this today and learned a lot -- a lot of the api doesn't work like it should appear to on the surface. Apologies for the confusion, as this also was not obvious to me.

It turns out that the are two different kinds of rules: system rules and "normal" rules.
_system_ rules are automatically created by the system, one for each detection list type / rule type combination. In the UI, these are labeled with a blue oval with the detection list name inside them.

The following api methods are only relevant to _system_ rules, and are designed to only be called by Code42-owner service accounts:
- add-users
- remove-users
- remove-all-users
- remove-user-aliases

Because users are not typically allowed to modify system rules, modifying them will not be part of this feature.

Unfortunately, the interface for modifying "normal" (user-defined) rules is all contained within a single `POST` for each api type:
- /api/v1/Rules/update-cloud-share-permissions-rule
- /api/v1/Rules/update-endpoint-exfiltration-rule
- /api/v1/Rules/update-file-type-mismatch-rule

Adding / removing users from a rule involves setting the exact desired list of users on the "userList" property in the post body: This example uses `cloud-share-permissions` rules but is relevant to all of them. You can observe this behavior by creating your own rule in the console and then modifying it.

```json
{
  "rule": {
    "tenantId": "string",
    "name": "string",
    "description": "string",
    "severity": 0,
    "isEnabled": true,
    "fileBelongsTo": {
      "usersToAlertOn": 0,
      "userList": [ // <--- this list needs to be modified
        "string"
      ]
    },
    "notificationConfig": {
      "enabled": true,
      "notificationInfo": [
        {
          "notificationType": 0,
          "notificationAddress": "string"
        }
      ]
    },
    "fileCategoryWatch": {
      "watchAllFiles": true,
      "fileCategoryList": [
        0
      ]
    },
    "ruleSource": "string",
    "watchGoogleDrive": {
      "publicOnTheWeb": true,
      "publicViaLink": true,
      "outsideTrustedDomains": true
    },
    "watchMicrosoftOneDrive": {
      "publicViaLink": true,
      "outsideTrustedDomains": true
    },
    "watchBox": {
      "publicViaLink": true,
      "outsideTrustedDomains": true
    },
    "id": "string"
  }
```

essentially, our `add_user` method will need to first call the corresponding `query` method for the correct rule type, then modify the "userList" on it, then `POST` the entire body. Fortunately, the response from `query` is mostly similar to what `update` expects, except for some metadata near the bottom:

```json
    {
      "tenantId": "string",
      "name": "string",
      "description": "string",
      "severity": 0,
      "isEnabled": true,
      "fileBelongsTo": {
        "usersToAlertOn": 0,
        "userList": [
          "string"
        ]
      },
      "notificationConfig": {
        "enabled": true,
        "notificationInfo": [
          {
            "notificationType": 0,
            "notificationAddress": "string"
          }
        ]
      },
      "fileCategoryWatch": {
        "watchAllFiles": true,
        "fileCategoryList": [
          0
        ]
      },
      "ruleSource": "string",
      "watchGoogleDrive": {
        "publicOnTheWeb": true,
        "publicViaLink": true,
        "outsideTrustedDomains": true
      },
      "watchMicrosoftOneDrive": {
        "publicViaLink": true,
        "outsideTrustedDomains": true
      },
      "watchBox": {
        "publicViaLink": true,
        "outsideTrustedDomains": true
      },
      "id": "string",
      // --- everything below this line should be removed before posting to update
      "createdAt": "2020-04-27T22:18:48.007Z",
      "createdBy": "string",
      "modifiedAt": "2020-04-27T22:18:48.007Z",
      "modifiedBy": "string",
      "isSystem": true
    }
```